onnxruntime/core/mlas/lib/amd64/cvtfp16Avx2.asm

```diff
@@ -0,0 +1,143 @@
+;++
+;
+; Copyright (c) Intel Corporation. All rights reserved.
+;
+; Licensed under the MIT License.
+;
+; Module Name:
+;
+;   cvtfp16Avx2.asm
+;
+; Abstract:
+;
+;   This module implements routines to convert between FP16 and FP32 formats using the AVX_NE_CONVERT ISA.
+;
+;--
+
+        .xlist
+INCLUDE mlasi.inc
+        .list
+
+        SUBTTL  "Convert buffer of half-precision floats to single-precision floats"
+;++
+;
+; Routine Description:
+;
+;   This routine converts the source buffer of half-precision floats to the
+;   destination buffer of single-precision floats.
+;
+;   This implementation uses AVX2 instructions.
+;
+; Arguments:
+;
+;   Source (rcx) - Supplies the address of the source buffer of half-precision
+;       floats.
+;
+;   Destination (rdx) - Supplies the address of the destination buffer of
+;       single-precision floats.
+;
+;   Count (r8) - Supplies the number of elements to convert.
+;
+; Return Value:
+;
+;   None.
+;
+;--
+
+        LEAF_ENTRY MlasConvertHalfToFloatBufferAVX2, _TEXT
+
+        test    r8, r8
+        jz      ExitRoutine
+        cmp     r8, 8
+        jb      ConvertMaskedVectors
+        cmp     r8, 16
+        jb      Convert128Vectors
+
+
+
+Convert256Vectors:
```

Comment:
This only uses 3 ymm registers when we have 16 ymm registers available. Do you know if there is any performance advantage gained buy using more ymm registers. in theory you could process 64 hfloats at a time using 6 ymm registers.

Reply:
the vcvtneeph2ps and vcvtneoph2ps have a latency of 7 and throughput of .333 (possible gain by loading more than just the 2 registers)
vunpckhps have latency of 1 and throughput of 1
vperm2f128 has a latency of 3 throughput of 1
vmovups has a latency of 7 throughput of 0.5